Change unapproved and unconfirmed account to not be accessible in the REST API

ClearlyClaire opened the pull request:
- [x] return 404 on `/api/v1/accounts` endpoints
- [x] change it so unapproved and unconfirmed accounts do not show up in search
- [x] do not allow mentioning unapproved and unconfirmed accounts
- [x] add tests

ClearlyClaire committed:
Change unapproved and unconfirmed account to not be accessible in the REST API

ClearlyClaire committed:
Change Account#searchable? to reject unconfirmed and unapproved users

ClearlyClaire committed:
Disable search for unapproved and unconfirmed users in Account.search_for

ClearlyClaire committed:
Disable search for unapproved and unconfirmed users in Account.advanced_search_for

ClearlyClaire committed:
Remove unconfirmed and unapproved accounts from Account.searchable scope

ClearlyClaire committed:
Prevent mentions to unapproved/unconfirmed accounts

ClearlyClaire commented:
I think the implementation itself is done but the queries still need to be reviewed for soundness (and tests need to be written) and performance impact.

ClearlyClaire committed:
Fix some old tests for Account.advanced_search_for

ClearlyClaire committed:
Add some Account.advanced_search_for tests for existing behaviors

ClearlyClaire committed:
Add some tests for Account.search_for

ClearlyClaire committed:
Add Account.advanced_search_for tests unconfirmed and unapproved accounts

ClearlyClaire committed:
Add Account.searchable tests

ClearlyClaire committed:
Fix Account.without_unapproved scope potentially messing with previously-applied scopes

ClearlyClaire commented:
Tests have been written, and the queries seem to work as expected. I have not fully reviewed the impact on performances yet, though.

ClearlyClaire commented:
I am unable to test performances on a large instance, but on a small instance with only a handful local users (but thousands of known remote ones), the performance impact is not observable.

Gargron commented on app/controllers/api/v1/accounts/lookup_controller.rb:
```diff
@@ -3,6 +3,8 @@
 class Api::V1::Accounts::LookupController < Api::BaseController
   before_action -> { authorize_if_got_token! :read, :'read:accounts' }
   before_action :set_account
+  before_action :check_account_approval
```
These changes would mean this API is no longer a reliable way to validate username availability on sign-up

ClearlyClaire replied:
Oh, I overlooked that use case for that API. I'm going to change that.

ClearlyClaire committed:
Allow lookup of unconfirmed/unapproved accounts through /api/v1/accounts/lookup

This is so that the API can still be used to check whether an username is free
to use.
